Add state protos and build script

dplumb94 opened the pull request:
To run, navigate to the root directory and `docker-compose up`. This will run the `processor/build.rs` script to build protos in the `processor/src/protobuf` directory

Commit:
Add state protos

Creates the state protos for bond in the sawtooth-bond/protos dir

Signed-off-by: Darian Plumb <dplumb@bitwise.io>

Commit:
Create initial docker-compose file

This docker-compose file will be used to run the whole bond app. For
now it just starts up a bond-processor container and builds the protos.

Signed-off-by: Darian Plumb <dplumb@bitwise.io>

dplumb94 commented:
I am wondering if the protos should be generated for each component as implemented here, or if there should be a script that is run from a shell container that puts the protos in a `sawtooth-bond/protobuf` directory, which could be required by each component. Any thoughts?

Update: resolved to create a common package and put protos there

eloaverona commented on protos/settlement.proto:
```diff
@@ -0,0 +1,45 @@
+    uint64 bond_quantity = 5;
+
+    // The amount of cash that is transferred
+    float currency_amount = 6;
```
As per @peterschwarz comment on Slack, I imagine currency amounts should not be floats in the protos either. 

> Currency amount should be an integer, probably millions of a single unit
e.g. $1 == 1,000,000

eloaverona commented on processor/Cargo.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
```
Does this need the copyright notice as well?

dplumb94 replied:
added

eloaverona commented on protos/bond.proto:
```diff
@@ -0,0 +1,69 @@
+   // Format: MM/DD/YYYY
+   string first_settlement_date = 11;
+}
+
```
Just to confirm, was it on purpose that we went away with the field [`coupon-benchmark`](https://sawtooth.hyperledger.org/docs/core/releases/0.7/bond_family_guide/design/object_model/bond.html#attributes)?  https://sawtooth.hyperledger.org/docs/core/releases/0.7/bond_family_guide/design/object_model/bond.html#attributes

shannynalayna replied:
Our intention with not including the coupon-benchmark parameter was based on the fact that the attribute, if set, must be set to 'Libor' and is only required if the bond has a floating coupon type. Therefore, to us, it seemed unnecessary.

eloaverona replied:
Makes sense! I just wanted to double-check it was intentional and not a forgotten field.

eloaverona commented on protos/organization.proto:
```diff
@@ -0,0 +1,70 @@
+    string pricing_source = 6;
+
+    // Map of participants to roles
+    map<string, Role> authorization = 7;
```
Would it make sense to have authorization be similar to holdings, as in a repeated field? I think it might increase clarity.

dplumb94 replied:
That sounds like a good idea. I'll update it

eloaverona commented on protos/quote.proto:
```diff
@@ -0,0 +1,62 @@
+    string quote_id = 1;
+
+    // Ticker symbol of the organization the quote is on bahalf of
+    string organization = 2;
```
Can we name this `organization_id` as well? To keep it consistent across protos

dplumb94 replied:
Right. We should make sure to reflect that in the transactions and DB as well.

eloaverona replied:
Can you rename this?

eloaverona commented on protos/settlement.proto:
```diff
@@ -0,0 +1,45 @@
+    string order_id = 1;
+
+    // Ticker symbol of the organization on whose behalf the order was entered
+    string ordering_organization = 2;
```
Name it `ordering_organization_id`? Again, for consistency.

eloaverona commented on protos/receipt.proto:
```diff
@@ -0,0 +1,47 @@
+    string bond_id = 1;
+
+    // Ticker symbol of the organization to which the payment was paid
+    string payee = 2;
```
Maybe this could be named `organization_id` or `payee_organization_id`? To keep it consistent across protos, and make it clear they are all referring to the same field in organization.

shannynalayna commented on protos/settlement.proto:
```diff
@@ -0,0 +1,46 @@
+
+    // Amount of cash that is transferred.
+    // Format: Millionths of dollars
+    float currency_amount = 6;
```
Same question as above, changing currency type from float to int?

shannynalayna commented on protos/organization.proto:
```diff
@@ -0,0 +1,108 @@
+
+        // Amount received.
+        // Format: Millionths of dollars
+        float amount = 4;
```
This currency value is also still a float type

shannynalayna commented on protos/organization.proto:
```diff
@@ -0,0 +1,108 @@
+
+        // The current balance of the holding.
+        // Format: Millionths of dollars, or quantity of bonds
+        float amount = 3;
```
Same as previous comment, were we converting these values to ints?

shannynalayna commented on protos/order.proto:
```diff
@@ -0,0 +1,89 @@
+    // not be set. If order_type is LIMIT, one of limit_price or limit_yield
+    // must be set.
+    // Format: Millionths of dollars
+    float limit_price = 6;
```
Are we changing the values formatted as millionths of dollars to ints?

dplumb94 replied:
Good catch! updated

shannynalayna reviewed (commented):
Some float types are left over and I thought we were switching those currency values to ints to represent millionths of dollars

Commit:
Add proto build script and add to common crate

- Creates a bond-common crate, where common functionality for
sawtooth-bond components can live.

- Adds proto generation functionality to build.rs, allowing other
components to use bond protos

- Updates .gitignore to ignore cargo build artifacts.

Signed-off-by: Darian Plumb <dplumb@bitwise.io>

Commit:
Initialize transaction processor

Creates the framework for the processor, including a Dockerfile.
Currently contains no functionality besides testing proto generation.

Signed-off-by: Darian Plumb <dplumb@bitwise.io>

shannynalayna commented on protos/bond.proto:
```diff
@@ -0,0 +1,77 @@
+
+   // Date on which the bond becomes worth it’s face value and can be redeemed.
+   // Format: UTC Timestamp
+   string maturity_date = 8;
```
Found another string that I believe should be an int
